src/core/or/circuituse.c

```diff
@@ -840,10 +840,15 @@ circuit_expire_building(void)
                  -1);
 
     circuit_log_path(LOG_INFO,LD_CIRC,TO_ORIGIN_CIRCUIT(victim));
-    if (victim->purpose == CIRCUIT_PURPOSE_C_MEASURE_TIMEOUT)
+    if (victim->purpose == CIRCUIT_PURPOSE_C_MEASURE_TIMEOUT) {
       circuit_mark_for_close(victim, END_CIRC_REASON_MEASUREMENT_EXPIRED);
-    else
+    } else {
       circuit_mark_for_close(victim, END_CIRC_REASON_TIMEOUT);
+      if (victim->purpose >= CIRCUIT_PURPOSE_S_HS_MIN_ &&
```

Comment:
Given that `hs_service_circuit_timed_out()` is pseudo-generic I think we might want to include `CIRCUIT_PURPOSE_HS_VANGUARDS` here, like `circuit_purpose_is_hidden_service()` does. Perhaps we can make a new function for this check, or add an `include_client` argument to `circuit_purpose_is_hidden_service()`? No strong opinions here.

Reply:
Oh good idea. Forgot about the existence of `circuit_purpose_is_hidden_service()`

Fixup: 986e82279d

Reply:
I'm not sure if that's the right function to use since it also includes client-side onion service circuits. That's why I suggested the modifications above.

Reply:
Oh damn... good catch. No it shouldn't be for client in this case at all. This is for "service" only.

I've removed entirely that fixup ^ and added a new fresh commit that splits up the `is_hidden_service()` into client and service specific functions. WAY LESS confusing that way imo.

See commit: 7eedb90850